Implementation of group creation function

BayG opened the pull request:
# WHAT
Implementing group creation function on this ChatSpace.

# WHY
To be capable of adding new chat group.

BayG committed:
make pull request

BayG committed:
Generated a controller file of groups

BayG committed:
[Fix] Changed the way of positioning element 'edit-group'

BayG committed:
make pull request

BayG committed:
Cloned view files already prepared, then replaced those to old files.

ssss- -

BayG committed:
[Fix] Changed file name; _user.scss → users.scss

BayG committed:
Added actions on method resources

BayG committed:
Add method new and edit on groups_controller.rb to enable to display view files

BayG committed:
[Fix] Edited the background-color of main-content-body

BayG committed:
[Fix] Improved  texts for describing view correctly

BayG committed:
Installed debugging tool: Binding-pry

BayG committed:
Reorganized controller: messages

-  Define creation of instance of class MessagesController in index
action
- Defined create action (still no valid contents)

BayG committed:
Reorganized controller: groups

-  Defined Group.all in action index.
- Defined action new:
 = Creation of an instance of class GroupsController.
 = have set current_user
- action edit (empty)
- Defined action create:
 = Have set conditional branch
- Private method: defined group_params

BayG committed:
Described Association: message

Notable: Inserted validation

BayG committed:
Described Association: group

Notable points:
 - Inserted validation
 - Inserted accepts_nested_attributes_for (For saving and updating all
at once including the model of the association destination.)

BayG committed:
changed from HTML notation to Ruby notation

BayG committed:
Reconfigure routing settings

BayG commented:
所々コメントアウトありますが、それは後の進捗に応じて有効にするつもりのものです。

sato11 commented on Gemfile:
```diff
@@ -34,6 +34,7 @@ gem 'haml'
 gem "haml-rails", "~> 0.9"
 gem "font-awesome-rails"
 gem 'devise'
+gem 'pry-rails'
```
このgemは本番環境で動作しないようにしましょう！

BayG replied:
group :developmentに入れればいいですね
[追記]group :development, :test に入れました。

sato11 commented on app/assets/stylesheets/_reset.scss:
```diff
@@ -10,7 +10,7 @@
     html{
         color:#000;
-        background:#FFF;
+        background: rgba(246, 246, 246, 0.5);
```
質問なのですが、こちらはリセットCSSに設定すべき値でしょうか？

BayG replied:
指摘されたので改めて考えてみましたが、その必要なさそうです。記述を初期に戻し、この値を_variable.scssに設定し直します。

sato11 commented on app/assets/stylesheets/messages.scss:
```diff
@@ -36,16 +39,13 @@ li {
+        }
+        .edit-account-btn {
+          float: right;
+          margin: 0 0 0 10px;
```
大きな理由がなければ margin-left で指定しましょう。

BayG replied:
修正しました。

sato11 commented on app/controllers/groups_controller.rb:
```diff
@@ -0,0 +1,28 @@
+    if @group.save
+      redirect_to root_path
+    else
+      render :new
```
成功時と失敗時にそれぞれフラッシュメッセージが出るようにいたしましょう。

BayG replied:
やってみます。

BayG replied:
groups_controller.rb 18行目と21行目にその機能を記述しました。

sato11 commented on app/views/groups/new.html.haml:
```diff
@@ -0,0 +1,27 @@
+.chat-group-form
+  %h1 新規チャットグループ
+  = form_for [@group] do |f|
```
new.html と edit.html は似通っているように見受けられますが、部分テンプレートにまとめられないでしょうか？

BayG replied:
やってみます。

sato11 commented on app/views/messages/index.html.haml:
```diff
@@ -5,7 +5,7 @@
           %p sampleAccount
         .sidebar-account-info__actionbtn
-          %a.sidebar-account-info__actionbtn--create-group
+          = link_to new_group_path, method: :get, class: 'create-group-btn' do
```
`method: :get` は省略しましょう。

BayG replied:
２箇所確認。省略しました。

sato11 commented on config/routes.rb:
```diff
@@ -1,5 +1,7 @@
   devise_for :users
   root      'messages#index'
-  resources :messages, only: [:index]
+  resources :groups, only: [:index, :new, :edit, :create] do
```
except の方が短くかけますね。

BayG replied:
やってみます。

BayG replied:
resources :groups, except: [:show, :destroy]
にしました。

sato11 commented on config/routes.rb:
```diff
@@ -1,5 +1,7 @@
 Rails.application.routes.draw do
   devise_for :users
   root      'messages#index'
```
こちらは変更なしでしょうか？

BayG replied:
root  'devise/sessions#new' で良いでしょうか

BayG replied:
失礼、groups#indexでしたね。

BayG replied:
直しました。

sato11 commented on app/controllers/groups_controller.rb:
```diff
@@ -0,0 +1,28 @@
+
+  def new
+    @group = Group.new
+    @user = current_user
```
この変数は何のために必要ですか？

BayG replied:
現在ログインしているユーザーのレコードをインスタンスとして使えるようにするためです。

sato11 commented on app/controllers/groups_controller.rb:
```diff
@@ -0,0 +1,28 @@
+class GroupsController < ApplicationController
+
+  def index
+    @groups = Group.all
```
ユーザー機能は実装済みかと思いますので、必要なグループのみ取得するようにしましょう。

BayG replied:
current_user.groups.new に置き換えました。

sato11 reviewed (commented):
"Implementation of group creation function" というタイトルですが、実際の作業範囲としてはどの程度を見込んでおられるでしょうか？

BayG commented:
<img width="960" alt="chat-space_create-groups" src="https://cloud.githubusercontent.com/assets/21117182/24092682/08577934-0d93-11e7-8b2d-87b56f9aa800.png">
Groupsテーブルにレコードが追加できました。

BayG committed:
put pry-rails onto group :development, :test

BayG committed:
[Fix] made a code for creating group record of current_user

BayG commented:
一つ前のプルリクの内容及びグループ編集画面の内容が含まれておりますが、基本タイトル通りの作業内容であると受け取って間違いありません。

BayG commented on app/controllers/groups_controller.rb:
```diff
@@ -0,0 +1,28 @@
+class GroupsController < ApplicationController
+
+  def index
+    @groups = Group.all
```
current_user.groups.new　を追記しました。

BayG replied:
現在このようなエラーが出てきました。原因解明に努めます。
<img width="1006" alt="chat-space_groups new_nomethoderror" src="https://cloud.githubusercontent.com/assets/21117182/24133456/6ad8556c-0e40-11e7-9302-2f5a05cba273.png">

BayG commented on app/controllers/groups_controller.rb:
```diff
@@ -0,0 +1,28 @@
+
+  def new
+    @group = Group.new
+    @user = current_user
```
上述の通りcurrent_user.groups.new　に置き換えました。

BayG committed:
[Fix] Edited .scss files

BayG committed:
[Fix] editing and adding Flash onto groups_controller.rb

- Changed "@groups = Group.all" to "current_user.groups.new".
- Added Flash messages function on conditional branch about editing a
chat group.

BayG committed:
[Fix] moved 'pry-rails' from group :development to "group :development, :test".

BayG committed:
[Fix] Fixed two places

- 'messages#index' → 'groups#index'
- only → except: [:show, :destroy]

BayG committed:
[Fix] Minor redundancy

Deleted method: :get cuz it is unnecessary to be written due to
compatibility for link_to method.

BayG committed:
Have used partial template

BayG commented:
<img width="1006" alt="chat-space_groups new_nomethoderror" src="https://cloud.githubusercontent.com/assets/21117182/24135115/9c446612-0e4b-11e7-9019-da580de76989.png">
現在このようなエラーが起こっており、解決に努めております。

【追記】解決しました。groups_controller.rb

BayG committed:
Created index view file of groups (rendered from messages/index.html.haml)

BayG committed:
[Fix] Fixed indentation

BayG committed:
Created flash message on conditional branch

BayG committed:
Added a description that allows the key of the flash message

kenchan0503 commented on app/controllers/groups_controller.rb:
```diff
@@ -0,0 +1,30 @@
+  end
+
+  def edit
+    @group = current_user.groups.find(group_params)
```
group_paramsの返り値の型とfindメソッドの引数の型が違いませんか？

BayG replied:
確認してみます。

BayG replied:
- findメソッドの引数は (:name) に変更
- group_name => name にカラム名を変更
- group_nameだった部分を一部を除きnameに記述を修正
以上３点を行いました。

kenchan0503 commented on app/controllers/groups_controller.rb:
```diff
@@ -0,0 +1,30 @@
+    if @group.save
+      redirect_to root_path, info: "チャットグルーブが作成されました"
+    else
+      render :new, danger: "グループ作成に失敗しました。"
```
groupが保存されなかった場合を指すものとして'danger'は少し大げさかなと思うのですがいかがでしょうか？

BayG replied:
確かにその通りですね。warningでどうでしょう。

BayG replied:
失礼、warningにせずerrorにすべて書き換えました。

kenchan0503 commented on app/controllers/messages_controller.rb:
```diff
@@ -1,6 +1,11 @@
 
   def index
+    @message = Message.new
+    # @group = Group.find(params[:group_id])
```
不要なコメントアウトは削除しましょう

kenchan0503 replied:
必要でしたら残しておいて構いません

BayG replied:
削除しました。

kenchan0503 commented on app/controllers/messages_controller.rb:
```diff
@@ -1,6 +1,11 @@
   end
 
+  def create
+    # Message.create(body: params[:body], image: params[:image], user_id: params[:user_id], group_id: params[:group_id])
```
こちらも

BayG replied:
同じく削除済みです。

kenchan0503 commented on app/models/group.rb:
```diff
@@ -1,2 +1,8 @@
 class Group < ApplicationRecord
+  validates :group_name, presence: true
```
groupテーブルのnameがgroupのnameであることは自明なのでgroup_nameよりnameの方がよいですね

BayG replied:
修正してみます。

kenchan0503 replied:
こちら修正は完了しておりますでしょうか？

kenchan0503 replied:
こちらのご確認もお願いいたします

BayG replied:
修正しました。

kenchan0503 replied:
修正されていないようです。
プッシュされましたでしょうか？

kenchan0503 commented on app/views/groups/_group.html.haml:
```diff
@@ -0,0 +1,25 @@
+    .chat-group-form__field--right
+      #chat-group-users
+        #chat-group-user-22.chat-group-user.clearfix
+          %input{:name => "chat_group[user_ids][]", :type => "hidden", :value => "22"}/
```
ハッシュロケット記法は避けましょう

BayG replied:
修正します。

BayG replied:
修正しました。

BayG committed:
[Fix] "danger" → "error"

BayG committed:
Deleted comment-out

kenchan0503 commented on app/controllers/groups_controller.rb:
```diff
@@ -0,0 +1,30 @@
+    if @group.save
+      redirect_to root_path, info: "チャットグルーブが作成されました"
+    else
+      render :new, error: "グループ作成に失敗しました。"
```
ここ、動いてますでしょうか？
動作確認をお願いいたします。お手数ですが該当画面のスクショもお願いいたします。

BayG replied:
<img width="1007" alt="chat-space_groups index" src="https://cloud.githubusercontent.com/assets/21117182/24144817/08885b6e-0e72-11e7-9961-9e9f125debec.png">
<img width="1107" alt="chat-space_group index_alert-info" src="https://cloud.githubusercontent.com/assets/21117182/24144821/0b3d82ee-0e72-11e7-91e6-defbd326a2fb.png">

BayG replied:
上：トップページ groups#index
下：チャットグループ追加が成功した後

kenchan0503 replied:
失敗時の画面もお願いいたします

kenchan0503 replied:
こちらもお願いいたします

kenchan0503 commented on app/controllers/groups_controller.rb:
```diff
@@ -0,0 +1,30 @@
+    end
+  end
+
+  def update
```
updateは後ほど実装されますか？

BayG replied:
そうします。後ほどグループ編集機能の実装で行うつもりです。

BayG commented:
<img width="678" alt="chat-space_groups-table_create-groups" src="https://cloud.githubusercontent.com/assets/21117182/24144933/88172a86-0e72-11e7-881d-ae2e7ffbc79f.png">
groupsテーブルにはこのように追加されたレコードがあります。

kenchan0503 commented on app/views/messages/index.html.haml:
```diff
@@ -62,3 +62,9 @@
         %label.main-content-footer__insert-file
           %i.fa.fa-image
         %input.main-content-footer.send-button{type: "submit", value: "Send"}
+      / = form_for [@group, @message] do |f|
```
不要なコメントアウトは消しましょう
必要でしたら残しておいて大丈夫です

BayG replied:
はい、必要です。メッセージ投稿機能実装時にコメントアウトを解くつもりです。
現在はエラー回避のためコメントアウト状態にしております。

BayG commented:
<img width="697" alt="chat-space_groups-table_name-column" src="https://cloud.githubusercontent.com/assets/21117182/24145812/43f360a0-0e76-11e7-8742-04b9a915fed7.png">
group_name　→　name　カラム名の変更が完了しました。

BayG committed:
[Fix] groups table: group_name → name column

BayG committed:
[Fix] Added flash message for error certainly.

BayG commented:
<img width="990" alt="chat-space_group index_alert-error" src="https://cloud.githubusercontent.com/assets/21117182/24147502/e3bfcbda-0e7d-11e7-972b-321ca66ae2d9.png">
グループ名の入力欄が空欄のときにSendボタンを押すと、このようにエラー用のflashメッセージが出るようにしました。

tetsu8dayo commented on app/views/layouts/application.html.haml:
```diff
@@ -9,6 +9,6 @@
     - flash.each do |key, value|
       - key = "info" if key == "notice"
-      - key = "danger"  if key == "alert"
+      - key = "error"  if key == "alert"
```
こちらでkeyの再代入を行っているのはどういう意図でしょうか？

tetsu8dayo replied:
私はadd_flash_typesを利用したことがないのですが、コードを見る限りadd_flash_typesの挙動を誤解していると思われます。（私が間違っていたら指摘してください。）
http://api.rubyonrails.org/classes/ActionController/Flash/ClassMethods.html#method-i-add_flash_types
こちらで確認しましょう！
このコードですと、特にadd_flash_typesをする意味はなさそうです。
https://github.com/rails/rails/blob/1413ee991ccfc76b24f29eb03c9cff82e588e5d7/actionpack/lib/action_controller/metal/flash.rb#L31

BayG replied:
参照URL：http://qiita.com/Yama-to/items/4d19a714d8bf5bfbabdd
結論から言うと、tetsuさんの指摘通りで概ね正しいと思います。
この方法は、最大で４種類のflashメッセージを実装できるための記述でしたが、あくまで今回は成功か警告かの２種類なので、以上の方法は冗長でした。引用してくださった情報は参考になりました。これから修正します。

BayG replied:
修正完了致しました。

tetsu8dayo commented on app/controllers/groups_controller.rb:
```diff
@@ -0,0 +1,31 @@
+    if @group.save
+      redirect_to root_path, info: "チャットグルーブが作成されました"
+    else
+      flash.now[:error] = "グループ作成に失敗しました"
```
flash[:key]という書き方をするならadd_flash_typesをする意味がないと思いますがどうでしょう？
下の方のレビューに書いてあるので確認してください！

tetsu8dayo commented on db/migrate/20170321112908_rename_group_name_column_to_groups.rb:
```diff
@@ -0,0 +1,5 @@
+class RenameGroupNameColumnToGroups < ActiveRecord::Migration[5.0]
+  def change
+    rename_column :groups, :group_name, :name
+  end
```
create_groups_tableのマイグレーションをrollbackしてからやり直すとファイルの数が増えないのでより良かったと思います

BayG committed:
[Fix] Reorganized the method flash
